internal/client/config.go

```diff
@@ -56,6 +70,74 @@ type AuthInfo struct {
 	ClientKeyData []byte `json:"client-key-data,omitempty" datapolicy:"security-key"`
 }
 
+func (c *Config) Equal(c2 *Config) bool {
+	if c == c2 {
+		return true
+	}
+	if c == nil || c2 == nil {
+		return false
+	}
+	return c.Service.Equal(&c2.Service) && c.AuthInfo.Equal(&c2.AuthInfo)
+}
+
+func (s *Service) Equal(s2 *Service) bool {
+	if s == s2 {
+		return true
+	}
+	if s == nil || s2 == nil {
+		return false
+	}
+	return s.Server == s2.Server && s.TLSServerName == s2.TLSServerName &&
+		s.CertificateAuthority == s2.CertificateAuthority &&
+		bytes.Equal(s.CertificateAuthorityData, s2.CertificateAuthorityData)
+}
+
+func (a *AuthInfo) Equal(a2 *AuthInfo) bool {
+	if a == a2 {
+		return true
+	}
+	if a == nil || a2 == nil {
+		return false
+	}
+	return a.ClientCertificate == a2.ClientCertificate && a.ClientKey == a2.ClientKey &&
+		bytes.Equal(a.ClientCertificateData, a2.ClientCertificateData) &&
+		bytes.Equal(a.ClientKeyData, a2.ClientKeyData)
+}
+
+func (c *Config) DeepCopy() *Config {
+	if c == nil {
+		return nil
+	}
+	return &Config{
+		Service:  *c.Service.DeepCopy(),
+		AuthInfo: *c.AuthInfo.DeepCopy(),
+	}
+}
+
+func (s *Service) DeepCopy() *Service {
+	if s == nil {
+		return nil
```

Comment:
If s is nil why not panic this would be a misuse of an object unless I am missing something. If we are going to return a nill object we should also return an error or at least document that it's possible.  But then again would s ever be nil in a sane case?

Reply:
The idea was to allow recursive copying of structs that may contain optional (thus nullable) fields. Of course, one could test every optional field beforehand, but that'd be quite verbose. Also, the method becomes chain-able when not returning an error.